File: Python/Algorithm/Algorithm I/542. 01 Matrix.py
import collections
from typing import List


class Solution:
    def isValidCell(self, ROWS: int, COLUMNS: int, row: int, column: int) -> bool:
        """
        Checks if a given cell is within the bounds of the matrix.

        Args:
            ROWS: The number of rows in the matrix.
            COLUMNS: The number of columns in the matrix.
            row: The row of the cell to check.
            column: The column of the cell to check.

        Returns:
            True if the cell is within the bounds of the matrix, False otherwise.

        Time Complexity: o(1) since we are only make a 4 comparisons.
        Space Complexity: o(1) since we are not using any extra space.
        """
        isValidRow = 0 <= row < ROWS
        isValidColumn = 0 <= column < COLUMNS

        return isValidRow and isValidColumn

    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        """
        Updates a matrix such that each cell contains the minimum number of moves required to reach a cell
        containing 0.

        Args:
            mat: A list of lists of integers, representing the matrix to update.

        Returns:
            A list of lists of integers, representing the updated matrix.

        Time Complexity: o(m * n) where m is the number of rows and n is the number of columns in the matrix.
        Space Complexity: o(m * n) where m is the number of rows and n is the number of columns in the matrix.
        """
        DIRECTIONS = [(-1, 0), (1, 0), (0, -1), (0, 1)]

        ROWS = len(mat)
        COLUMNS = len(mat[0])

        queue = collections.deque()
        for row in range(ROWS):
            for column in range(COLUMNS):
                if mat[row][column] == 0:
                    queue.append([row, column])

                else:
                    mat[row][column] = -1

        while queue and queue[0]:
            row, column = queue.popleft()

            for rowStep, columnStep in DIRECTIONS:
                newRow = row + rowStep
                newColumn = column + columnStep

                if not self.isValidCell(ROWS, COLUMNS, newRow, newColumn) or mat[newRow][newColumn] != -1:
                    continue

                mat[newRow][newColumn] = mat[row][column] + 1
                queue.append([newRow, newColumn])

        return mat

File: Java/Algorithm/Algorithm I/542. 01 Matrix.java
class Solution {

  /**
   * Checks if a cell is a valid cell within the given matrix.
   *
   * @param mat    The matrix to check cell validity within.
   * @param row    The row index of the cell.
   * @param column The column index of the cell.
   * @return True if the cell is valid, otherwise false.
   */
  private boolean isValidCell(int[][] mat, int row, int column) {
    int ROWS = mat.length;
    int COLUMNS = mat[0].length;

    boolean isRowValid = row >= 0 && row < ROWS;
    boolean isColumnValid = column >= 0 && column < COLUMNS;

    return isRowValid && isColumnValid;
  }

  /**
   * Computes the minimum distance from 1-valued cells to 0-valued cells in a binary matrix.
   *
   * @param mat The binary matrix where 0 represents empty cells and 1 represents obstacles.
   * @return A matrix representing the minimum distance from each 1-valued cell to the nearest 0-valued cell.
   */
  public int[][] updateMatrix(int[][] mat) {
    int ROWS = mat.length;
    int COLUMNS = mat[0].length;

    int[][] distanceMat = new int[ROWS][COLUMNS];
    Queue<int[]> queue = new LinkedList<>();

    for (int row = 0; row < ROWS; row++) {
      for (int column = 0; column < COLUMNS; column++) {
        if (mat[row][column] == 0) {
          queue.offer(new int[] { row, column });
          distanceMat[row][column] = 0;
        } 
        
        else {
          distanceMat[row][column] = Integer.MAX_VALUE;
        }
      }
    }

    int[][] directions = { { -1, 0 }, { 1, 0 }, { 0, -1 }, { 0, 1 } };

    while (!queue.isEmpty()) {
      int[] currentCell = queue.poll();
      int row = currentCell[0];
      int column = currentCell[1];

      for (int[] direction : directions) {
        int newRow = row + direction[0];
        int newColumn = column + direction[1];

        if (isValidCell(mat, newRow, newColumn)) {
          if (distanceMat[newRow][newColumn] > distanceMat[row][column] + 1) {
            distanceMat[newRow][newColumn] = distanceMat[row][column] + 1;
            queue.offer(new int[] { newRow, newColumn });
          }
        }
      }
    }

    return distanceMat;
  }
}
